libsolidity/analysis/TypeChecker.cpp

```diff
@@ -139,13 +139,19 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 			toString(arguments.size()) +
 			" were provided."
 		);
-	if (arguments.size() >= 1 && !type(*arguments.front())->isImplicitlyConvertibleTo(ArrayType::bytesMemory()))
-		m_errorReporter.typeError(
+
+	BoolResult details{false};
+	if (
+		arguments.size() >= 1 &&
+		!(details= type(*arguments.front())->isImplicitlyConvertibleTo(ArrayType::bytesMemory()))
```

Comment:
I think we should limit assignments inside expressions in boolean context to the outer layer of an if condition.